src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -526,6 +526,162 @@ private static int GetFloatingPointMaxDigitsAndPrecision(char fmt, ref int preci
             return maxDigits;
         }
 
+        private static void FormatFloatAsHex<TNumber, TChar>(ref ValueListBuilder<TChar> vlb, TNumber value, char fmt, int precision)
+            where TNumber : unmanaged, IBinaryFloatParseAndFormatInfo<TNumber>
+            where TChar : unmanaged, IUtfChar<TChar>
+        {
+            // Get the raw bits
+            ulong bits = TNumber.FloatToBits(value);
+            int mantissaBits = TNumber.NormalMantissaBits;
+            int exponentBias = TNumber.ExponentBias;
+
+            // Extract sign, exponent, and mantissa
+            bool isNegative = (bits >> (mantissaBits + TNumber.ExponentBits)) != 0;
+            int biasedExponent = (int)((bits >> mantissaBits) & ((1UL << TNumber.ExponentBits) - 1));
+            ulong mantissa = bits & TNumber.NormalMantissaMask;
+
+            // Add sign
+            if (isNegative)
+            {
+                vlb.Append(TChar.CastFrom('-'));
+            }
+
+            // Add "0x" prefix
+            vlb.Append(TChar.CastFrom('0'));
+            vlb.Append(TChar.CastFrom(fmt)); // 'x' or 'X'
+
+            // Handle special cases
+            if (biasedExponent == TNumber.InfinityExponent)
+            {
+                // Infinity or NaN - just output as 0
+                vlb.Append(TChar.CastFrom('0'));
+                vlb.Append(TChar.CastFrom('p'));
+                vlb.Append(TChar.CastFrom('+'));
+                vlb.Append(TChar.CastFrom('0'));
+                return;
+            }
+
+            if (biasedExponent == 0 && mantissa == 0)
+            {
+                // Zero
+                vlb.Append(TChar.CastFrom('0'));
+                if (precision > 0)
+                {
+                    vlb.Append(TChar.CastFrom('.'));
+                    for (int i = 0; i < precision; i++)
+                    {
+                        vlb.Append(TChar.CastFrom('0'));
+                    }
+                }
+                vlb.Append(TChar.CastFrom('p'));
+                vlb.Append(TChar.CastFrom('+'));
+                vlb.Append(TChar.CastFrom('0'));
+                return;
+            }
+
+            // Normalize mantissa for hex output (leading digit should be 1.xxx in range [1, 2))
+            int actualExponent;
+            ulong significand;
+
+            if (biasedExponent == 0)
+            {
+                // Denormal number - normalize by shifting until we get leading 1
+                significand = mantissa;
+                int lz = BitOperations.LeadingZeroCount(significand) - (64 - mantissaBits);
+                significand <<= lz;
+                actualExponent = 1 - exponentBias - lz;
+            }
+            else
+            {
+                // Normal number - add implicit leading bit
+                significand = (1UL << mantissaBits) | mantissa;
+                actualExponent = biasedExponent - exponentBias;
+            }
```

Comment:
We have a `DiyFp` type that already centralizes this logic and helps ensure it is correct over time. It should be used for getting the "normalized" data.